### PR TITLE
Ensure that for RWX volumes expansion is called on all nodes

### LINKED ISSUE #80918: CSI resize doesn't handle multi-attached volumes quite right
if a CSI volume is attached to more than 1 node, and it is expanded, and if the CSI plugin returns true from ControllerResizeVolume, then the PVC will get marked for node resize, but there's no mechanism to ensure that every node currently using the volume will call the CSI node plugin.

**What happened**: One node called the NodeExpandVolume RPC, the rest did not.

**What you expected to happen**: All nodes should call the NodeExpandVolume RPC.

**How to reproduce it (as minimally and precisely as possible)**: Use a CSI plugin that supports mutli-attach, create a PVC using that CSI plugin, create 2 or more pods using the same PVC on different nodes, then resize the PVC.

**Anything else we need to know?**: There are 2 necessary activities getting combined into a single operation: 1. node-side work required to complete the resize operation, such as resizing of a filesystem and 2. node-side work required to make pods aware of the updated volume size. (1) needs to be done only a single time but needs to be done before the resize can be considered complete. It needs to be done even if the volume is not in use at the time of the resize. (2) needs to be done for every node currently accessing the volume so they all find out about the new size, but if no nodes are accessing the volume, then it could be skipped without harm.

Consider the case of an iSCSI-based plugin implementing raw block volumes. iSCSI supports r/w multi attach as long as it's raw blocks and not filesystem mode. To perform a resize of such a volume, the underlying LUN needs to be resized (on the storage controller) and then the iSCSI initiators all have to rescan their buses to see the new size. In CSI, the iSCSI rescan operation needs to happen in NodeExpandVolume, because this is the only place where such a thing is possible. This is a case of using NodeExpandVolume for purpose (2) above, though, and kubelet only supports purpose (1) currently.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.15.1
- Cloud provider or hardware configuration: n/a
- OS (e.g: `cat /etc/os-release`): Arch Linux
- Kernel (e.g. `uname -a`): 5.2.5-arch1-1-ARCH
- Install tools: n/a
- Network plugin and version (if this is a network-related bug): bridge
- Others: containerd

### PR DESCRIPTION
**Pull request by gnufied**
Also if CSI driver throws volume-in-use error do not retry node expansion.

This PR adds desired size based on PV capacity to desired_state_of_the_world and actual size based on `pvc.Status.Capacity` to actual_state_of_the_world. Actual size is calculated the time when pod enters ASOW. This way if a RWX volume that is node-expanded on one node will not have actual PVC size updated on second node until second node also calls `NodeExpandVolume`. 

Fixes https://github.com/kubernetes/kubernetes/issues/80918 and https://github.com/kubernetes/kubernetes/issues/92931

/kind bug
/sig storage

```release-note
Do not retry NodeExpandVolume if CSI drivers throw volume-in-use error
Call NodeExpandVolume on all RWX nodes
```

### CODE REVIEW

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (msau42) [inline pkg/kubelet/volumemanager/cache/actual_state_of_world.go:413]: What happens if you expand a volume twice?

**Review by msau42 — COMMENTED** [review]
Can you explain how this change will call expand on all nodes?  What if by the time a Pod enters dsw one node2, node1 has already updated pvc status?

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (gnufied) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go:413]: if you expand a volume twice then only size recorded in DSOW changes until `NodeExpandVolume` is called and at which point size recorded in ASOW s updated as well.

**Comment by gnufied** [comment]
@msau42 updated the description with how it fixes RWX expansion.  This PR only fixes online expansion for RWX volumes if volume was offline on one node and online on another, then `NodeExpandVolume` will be called only on one node. I think that is reasonable behaviour.

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go
Comment (msau42) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:411]: this seems odd. We update asw in dswp? This is because reconstruction doesn't know/have the pvc status? CSI at least keeps a file, so we could use that?

File: pkg/volume/util/operationexecutor/operation_generator.go
Comment (msau42) [inline pkg/volume/util/operationexecutor/operation_generator.go:1666]: In the readonly case above, or any other case where we return true early, do we need to mark volume as resized? https://github.com/kubernetes/kubernetes/pull/92932/files#diff-450e811a4953f760ff1594ede8b2037eR1637

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (msau42) [inline pkg/kubelet/volumemanager/cache/actual_state_of_world.go:269]: Can you add some more comments here explaining how this is populated/updated, how it addresses the RWX case, and the known races?

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (msau42) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go:413]: Not related to this change, but it may be worth thinking of a scenario where someone increases the size twice in quick succession. Would it be possible that we resize the fs for the first expansion, but when we update the pvc status, it updates the size to the 2nd value? (maybe an e2e test could be valuable too)

File: pkg/volume/util/operationexecutor/operation_generator.go
Comment (gnufied) [reply pkg/volume/util/operationexecutor/operation_generator.go:1666]: we don't need to mark them as resized as such because such volumes are not even entered into DSOW's `persistentVolumeSize` field. 

for readonly volumes, we don't track their sizes in DSOW - https://github.com/kubernetes/kubernetes/pull/92932/files#diff-94e527d26427ceb4f990f05f76952656 and hence will not result in resize. I have also added an unit test that covers readonly volumes

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (gnufied) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go]: done.

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go
Comment (gnufied) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:411]: Yes, that is indeed the case. I would like to this more about recording pvc's status size in json file. Can I do this in a follow up?

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (gnufied) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go:416]: yes If someone increases the size in quick succession and assuming control-plane is finished doing its thing, then size of PV will be second value. But in this case when `NodeExpandVolume` is invoked, it will get 2nd value recorded in PV as newSize (and not the intermediate one) and hence `pvc.Status.Capacity` will be updated with 2nd value.

File: pkg/volume/util/operationexecutor/operation_generator.go
Comment (msau42) [reply pkg/volume/util/operationexecutor/operation_generator.go:1666]: Would we still need to update the pvc status to say resizing is finished?

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go
Comment (msau42) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:411]: sure can you open an issue to track

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (msau42) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go:416]: Do we need to think about a way we can address this? Such as recording the size before we call node expansion and using that value to update? This is kind of related to the challenge that the driver doesn't actually tell us the size.

File: pkg/volume/util/operationexecutor/operation_generator.go
Comment (gnufied) [reply pkg/volume/util/operationexecutor/operation_generator.go:1666]: we can't mark PVC resizing as finished because if PVC was readonly on this node, resizing will never happen whereas it can successfully happen on some other node.

Comment (gnufied) [reply pkg/volume/util/operationexecutor/operation_generator.go:1666]: Having said that - all the `return true, blah` statements look weird but we are basically double checking scenarios we already checking when a volume enters DSOW for volume expansion. May be that can be cleaned up but it does not hurt to be safe.

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go
Comment (gnufied) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:411]: https://github.com/kubernetes/kubernetes/issues/92947

File: pkg/controller/volume/attachdetach/attach_detach_controller.go
Comment (saad-ali) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go:380]: Something seems wrong here since you are having to pass new params to A/D controller but don't use those new params in A/D controller.

Comment (gnufied) [reply pkg/controller/volume/attachdetach/attach_detach_controller.go:380]: yeah I thought about it, it is somewhat ugly. we don't track `pvc.status.capacity` in control-plane for volume expansion in attach-detach controller but we do in volumemanager and hence this parameter is unused in A/D controller.

I could introduce a new interface function but it has to be part of `ActualXXUpdater` interface and A/D controller side of `ASOW` has to provide a dummy implementation.  I am happy to do that instead of this.

**Comment by msau42** [comment]
/hold
to address pending comment

**Comment by saad-ali** [comment]
Taking a deeper look

File: pkg/controller/volume/attachdetach/attach_detach_controller.go
Comment (gnufied) [reply pkg/controller/volume/attachdetach/attach_detach_controller.go:380]: I have a commit that introduces the new interface function called `MarkAsAttachedWithSize` inside `ActualStateOfWorldAttacherUpdater` interface.  I have not pushed it because I am afraid to make CI unhappy at last minute - https://github.com/gnufied/kubernetes/commit/983e410660fcec7bb0fb019f9d68eb4d7eaf333a#diff-24ba5bb87df94b8b68b6fbd13ef972b0R262 

But I am happy to do this in a follow up. The change is mostly cosmetic and does not changes underlying logic of the PR.

Comment (gnufied) [reply pkg/controller/volume/attachdetach/attach_detach_controller.go]: done

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (saad-ali) [inline pkg/kubelet/volumemanager/cache/actual_state_of_world.go:115]: Don't collapse - scope the input for each method down to the bare minimum it needs.

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (saad-ali) [inline pkg/kubelet/volumemanager/cache/actual_state_of_world.go:275]: This code is not CSI specific

File: pkg/volume/util/operationexecutor/operation_executor.go
Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_executor.go:224]: Move this to `type ActualStateOfWorld interface` in `kubernetes/kubernetes/pkg/kubelet/volumemanager/cache/actual_state_of_world.go`. The methods in `ActualStateOfWorldAttacherUpdater` are common for both A/D controller and kubelet volume manager. Since this only applies to the later we don't need to put it in the common interface.

File: pkg/volume/util/operationexecutor/operation_executor.go
Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_executor.go:208]: Move this to `type ActualStateOfWorld interface` in `kubernetes/kubernetes/pkg/kubelet/volumemanager/cache/actual_state_of_world.go`. The methods in `ActualStateOfWorldAttacherUpdater` are common for both A/D controller and kubelet volume manager. Since this only applies to the later we don't need to put it in the common interface.

**Review by saad-ali — CHANGES_REQUESTED** [review]

File: pkg/volume/util/operationexecutor/operation_executor.go
Comment (gnufied) [reply pkg/volume/util/operationexecutor/operation_executor.go:224]: We can't move this function into `ActualStateOfWorld` interface of `volumemanager/cache` because - it will cause cyclic dependency.

File: pkg/volume/util/operationexecutor/operation_executor.go
Comment (gnufied) [reply pkg/volume/util/operationexecutor/operation_executor.go:208]: same as above.

File: pkg/volume/util/operationexecutor/operation_executor.go
Comment (gnufied) [reply pkg/volume/util/operationexecutor/operation_executor.go:224]: see - cache.ActualStateWorld already depends on `operationexecutor` package, now if we say operationexecutor will depend on `cache.ActualStateWorld` it will cause cyclic dependency.

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (gnufied) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go:115]: In general I agree but if we keep each argument separate then this function will now need 4 arguments and IMO i would be happy if it were 3 but 4 is tad too much. 

I know in other places we have used https://refactoring.guru/introduce-parameter-object pattern when number of parameters to a function grow too big, for example `MarkVolumeAsMounted` will require 9 arguments if specified separately.

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go
Comment (gnufied) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go]: fixed

**Comment by nikhita** [comment]
👋 Hi, just checking if this is still planned for 1.19 and if this needs to go in before code freeze is lifted? Can this be cherry-picked to the `release-1.19` branch after code freeze is lifted?

**Comment by gnufied** [comment]
@nikhita this PR can be removed from 1.19 milestone. The important bug we wanted to fix has been merged via https://github.com/kubernetes/kubernetes/pull/92986 . I need to rework this PR little bit to what was discussed with @saad-ali offline.

**Comment by nikhita** [comment]
@gnufied Thanks for the update! I've updated the milestone to v1.20. :+1:

**Comment by sayanchowdhury** [comment]
👋🏽 Hello from Bug Triage team! The code freeze is starting November 12th (about 3 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.20, is it still planned for this release?

**Comment by droslean** [comment]
@gnufied Can you check this? Are you expecting that this PR will land in 1.20?

**Comment by droslean** [comment]
Since there is no activity on this PR for many days, and the code freeze has started, I am moving this to 1.21

/milestone v1.21

**Comment by shubheksha** [comment]
@gnufied do you plan to get this merged?

**Comment by erismaster** [comment]
Hi 👋 Bug Triage here, just a reminder that code freeze is end of day, tomorrow the 9th! PRs need to be merged in before then, or they will require an approved exception to go into the 1.21 milestone.

Also, this PR looks hasn't had much activity lately, as such if we don't hear any updates I will be dropping the milestone until it is active again. Thanks!

**Comment by erismaster** [comment]
This PR has not had much activity for a few releases now, I am going to clear the milestone. If this becomes active again, it can be added into the relevant milestone at that time. Thanks!

/milestone clear

**Pull request closed without merging**
